File: demo/c/solar_time.c
/*
    solar_time.c  -  by Don Cross - 2019-06-11

    Example C program for Astronomy Engine:
    https://github.com/cosinekitty/astronomy

    Given an observer's geographic latitude and longitude,
    and an optional date and time, this program displays
    true solar time for that observer and time.
*/

#include <stdio.h>
#include <math.h>
#include "astro_demo_common.h"

int main(int argc, const char *argv[])
{
    int error;
    astro_observer_t observer;
    astro_time_t time;
    astro_func_result_t ha;
    double solarTimeHours;
    int hour, minute, second, milli;

    error = ParseArgs(argc, argv, &observer, &time);
    if (error)
        return error;

    ha = Astronomy_HourAngle(BODY_SUN, &time, observer);
    if (ha.status != ASTRO_SUCCESS)
    {
        printf("ERROR %d in Astronomy_HourAngle().\n", ha.status);
        return 1;
    }

    solarTimeHours = fmod(ha.value + 12.0, 24.0);

    milli = (int)round(solarTimeHours * 3600000.0);
    second = milli / 1000;
    milli %= 1000;
    minute = second / 60;
    second %= 60;
    hour = minute / 60;
    minute %= 60;
    hour %= 24;

    printf("True solar time = %0.4lf hours (%02d:%02d:%02d.%03d)\n", solarTimeHours, hour, minute, second, milli);
    return 0;
}

File: demo/kotlin/src/main/kotlin/SolarTime.kt
import io.github.cosinekitty.astronomy.*
import kotlin.math.roundToInt

/**
 * Calculate true solar time for a given observer and UTC time.
 *
 * @param observer
 * The geographic location for which to calculate apparent celestial body positions.
 *
 * @param time
 * The date and time for which to calculate positions.
 */
internal fun `Solar true time`(observer: Observer, time: Time): Int {
    val ha = hourAngle(Body.Sun, time, observer)
    val solarTimeHours = (ha + 12.0) % 24.0
    var milli = (solarTimeHours * 3.6e+6).roundToInt()
    var second = milli / 1000
    milli %= 1000
    var minute = second / 60
    second %= 60
    var hour = minute / 60
    minute %= 60
    hour %= 24

    println("True solar time = %7.4f hours (%02d:%02d:%02d.%03d)".format(
        solarTimeHours, hour, minute, second, milli))

    return 0
}
